dali/pipeline/data/buffer_manager.h

```diff
@@ -0,0 +1,179 @@
+  if (current_idx == -1) {
+    if (std::is_same<Backend, GPUBackend>::value) {
+      std::cout << "Did not find the sufficient buffer." << std::endl;
+    }
+    auto b = new Buffer<Backend>;
+    b->set_pinned(pinned);
+    buff.reset(b);
+  } else {
+    // here we have the best fit - get the buffer and remove from vector
+    buff = std::move((*buffers)[current_idx]);
+    // remove the (now empty) buffer from the pool
+    // Note: For some reason using buffers.erase(buffers.begin() + current_idx)
+    // causes huge numbers of allocations..
+    std::swap((*buffers)[current_idx], buffers->back());
+    buffers->pop_back();
+  }
+
+  return buff;
+}
+
+class BufferManagerBase {
+ public:
+  explicit BufferManagerBase(int device) : device_(device) {}
+  ~BufferManagerBase() = default;
+
+  virtual unique_ptr<Buffer<CPUBackend>> AcquireBuffer(const size_t nbytes, const bool pinned) = 0;
+  virtual unique_ptr<Buffer<GPUBackend>> AcquireBuffer(const size_t nbytes) = 0;
+
+  virtual void ReleaseBuffer(unique_ptr<Buffer<CPUBackend>> *buffer, const bool pinned) = 0;
+  virtual void ReleaseBuffer(unique_ptr<Buffer<GPUBackend>> *buffer) = 0;
+
+ protected:
+  int device_;
+
+  // locks to ensure thread-safety of buffer lists
+  std::mutex cpu_buffers_mutex_, pinned_buffers_mutex_, gpu_buffers_mutex_;
+};
+
+class LinearBufferManager : public BufferManagerBase {
+ public:
+  explicit LinearBufferManager(int device) :
+    BufferManagerBase(device) {
+    std::lock_guard<std::mutex> lock(cpu_buffers_mutex_);
+    int original_device;
+    CUDA_CALL(cudaGetDevice(&original_device));
+    CUDA_CALL(cudaSetDevice(device_));
+
+    const int buffers_init_size = 0;
+    const size_t buffer_initial_size = 1048576;
+
+    cpu_buffers_.resize(buffers_init_size);
+    pinned_buffers_.resize(buffers_init_size);
+    gpu_buffers_.resize(buffers_init_size);
+
+    for (int i = 0; i < buffers_init_size; ++i) {
+      cpu_buffers_[i].reset(new Buffer<CPUBackend>);
+      cpu_buffers_[i]->Resize({buffer_initial_size});
+
+      pinned_buffers_[i].reset(new Buffer<CPUBackend>);
+      pinned_buffers_[i]->set_pinned(true);
+      pinned_buffers_[i]->Resize({buffer_initial_size});
+
+      gpu_buffers_[i].reset(new Buffer<GPUBackend>);
+      gpu_buffers_[i]->Resize({buffer_initial_size});
+    }
+
+    CUDA_CALL(cudaSetDevice(original_device));
+  }
+
+  // Acquire a buffer
+  inline unique_ptr<Buffer<CPUBackend>> AcquireBuffer(const size_t nbytes,
+                                                      const bool pinned) override;
+  inline unique_ptr<Buffer<GPUBackend>> AcquireBuffer(const size_t nbytes) override;
+
+  // Release a buffer
+  inline void ReleaseBuffer(unique_ptr<Buffer<CPUBackend>> *buffer,
+                            const bool pinned) override;
+  inline void ReleaseBuffer(unique_ptr<Buffer<GPUBackend>> *buffer) override;
+
+ private:
+  // Buffer lists for re-use
+  vector<unique_ptr<Buffer<CPUBackend>>> cpu_buffers_;
```

Comment:
Maybe list would be better than vector if we are doing only linear searches?

Reply:
I don't think it would be faster. Also I'm not 100% sure we are only going to do linear searches even though right now we do, so I would leave it as is.